pkg/httpd/webfolder.go

```diff
@@ -20,56 +20,67 @@ func (h *webFolder) Name() string {
 }
 
 func (h *webFolder) CheckValidation() error {
-	apiClient := h.ws.apiClient
-	user := h.ws.CurrentUser()
-	terminalCfg, err := h.ws.apiClient.GetTerminalConfig()
+	if volume, err := SftpCheckValidation(h.ws); err != nil {
+		return err
+	} else {
+		h.volume = volume
+	}
+	return nil
+}
+
+func (h *webFolder) HandleMessage(*Message) {
+}
+
+func (h *webFolder) CleanUp() {
+	close(h.done)
+	h.volume.Close()
+}
+
+func (h *webFolder) GetVolume() *UserVolume {
+	select {
+	case <-h.done:
+		return nil
+	default:
+		return h.volume
+	}
+}
+
+func SftpCheckValidation(ws *UserWebsocket) (*UserVolume, error) {
+	apiClient := ws.apiClient
+	user := ws.CurrentUser()
+	terminalCfg, err := ws.apiClient.GetTerminalConfig()
+
+	uv := &UserVolume{}
 	if err != nil {
 		logger.Errorf("Get terminal config failed: %s", err)
-		return err
+		return uv, err
 	}
 	volOpts := make([]VolumeOption, 0, 5)
 	volOpts = append(volOpts, WithUser(user))
-	volOpts = append(volOpts, WithAddr(h.ws.ClientIP()))
+	volOpts = append(volOpts, WithAddr(ws.ClientIP()))
 	volOpts = append(volOpts, WithTerminalCfg(&terminalCfg))
-	params := h.ws.wsParams
+	params := ws.wsParams
 	targetId := params.TargetId
 	assetId := params.AssetId
 	if assetId == "" {
 		assetId = targetId
 	}
-	if h.ws.ConnectToken != nil {
-		connectToken := h.ws.ConnectToken
+	if ws.ConnectToken != nil {
+		connectToken := ws.ConnectToken
 		volOpts = append(volOpts, WithConnectToken(connectToken))
 	} else {
 		if common.ValidUUIDString(assetId) {
 			assets, err := apiClient.GetUserAssetByID(user.ID, assetId)
 			if err != nil {
 				logger.Errorf("Get user asset %s error: %s", assetId, err)
-				return ErrAssetIdInvalid
+				return uv, ErrAssetIdInvalid
 			}
 			if len(assets) != 1 {
 				logger.Errorf("Get user more than one asset %s: choose first", targetId)
 			}
 			volOpts = append(volOpts, WithAsset(&assets[0]))
 		}
 	}
-	h.volume = NewUserVolume(apiClient, volOpts...)
-	return nil
-}
 
-func (h *webFolder) HandleMessage(*Message) {
-}
-
-func (h *webFolder) CleanUp() {
-	close(h.done)
-	h.volume.Close()
-}
-
-func (h *webFolder) GetVolume() *UserVolume {
-	select {
-	case <-h.done:
-		return nil
-	default:
-		return h.volume
-	}
+	return NewUserVolume(apiClient, volOpts...), nil
 }
```

Comment:
The given code has no syntax errors. In terms of optimization, we can avoid duplicating code if needed.
As the question does not contain enough specifics about specific concerns such as performance improvements or security enhancements required in this context, I cannot give further suggestions specifically related to these aspects.

Since the previous comment was generated from text that did not include actual function names or parameter values, it is unclear what "NewUserVolume" represents without additional information. As there are no explicit variable assignments with the word "volume", it seems unlikely that a new volume could be created directly within these functions based on existing structures.

Given the constraints provided, all possible changes would likely involve altering existing parameters and variables like `uv`, which might affect functionality slightly but wouldn't resolve the main concern regarding duplication of code execution. It's important to note that reworking complex interfaces may lead to unexpected side effects due to interactions between methods and arguments, which is why some optimizations should always be performed carefully and only when it makes sense.
In summary, while any adjustments you propose must consider the context and requirements, the lack of clarity around how `h.volume` relates to other elements suggests that making major structural variations will carry significant risks unless absolutely necessary.